File: tools/mergeNanoVDB.cpp
// ======================================================================== //
// Merge distributed NanoVDB files into a single grid
// Usage: ./mergeNanoVDB <input_dir> <output_file> [--timestep TIMESTEP]
// ======================================================================== //

#include <iostream>
#include <fstream>
#include <vector>
#include <string>
#include <filesystem>
#include <algorithm>
#include <cstring>

#ifdef HS_HAVE_NANOVDB
#include <nanovdb/NanoVDB.h>
#include <nanovdb/util/IO.h>
#include <nanovdb/util/GridBuilder.h>
#endif

namespace fs = std::filesystem;

struct Options {
  std::string inputDir;
  std::string outputFile;
  std::string timestep;
  std::string field;
  bool verbose = false;
};

void printUsage(const char* prog) {
  std::cout << "Usage: " << prog << " <input_dir> <output_file> [options]\n\n";
  std::cout << "Options:\n";
  std::cout << "  --timestep <TS>   Only merge files from specific timestep\n";
  std::cout << "  --field <name>    Only merge files for specific field\n";
  std::cout << "  --verbose, -v     Print detailed progress\n";
  std::cout << "\nExample:\n";
  std::cout << "  " << prog << " /data/nvdb merged_004320.nvdb --timestep 004320\n";
}

Options parseArgs(int argc, char** argv) {
  Options opts;
  
  if (argc < 3) {
    printUsage(argv[0]);
    exit(1);
  }
  
  opts.inputDir = argv[1];
  opts.outputFile = argv[2];
  
  for (int i = 3; i < argc; i++) {
    std::string arg = argv[i];
    if (arg == "--timestep" && i + 1 < argc) {
      opts.timestep = argv[++i];
    } else if (arg == "--field" && i + 1 < argc) {
      opts.field = argv[++i];
    } else if (arg == "--verbose" || arg == "-v") {
      opts.verbose = true;
    } else {
      std::cerr << "Unknown option: " << arg << std::endl;
      printUsage(argv[0]);
      exit(1);
    }
  }
  
  return opts;
}

std::vector<std::string> findNVDBFiles(const Options& opts) {
  std::vector<std::string> files;
  
  try {
    for (const auto& entry : fs::directory_iterator(opts.inputDir)) {
      if (entry.path().extension() != ".nvdb") continue;
      
      std::string filename = entry.path().filename().string();
      
      // Filter by timestep
      if (!opts.timestep.empty()) {
        if (filename.find("_" + opts.timestep + ".") == std::string::npos) {
          continue;
        }
      }
      
      // Filter by field
      if (!opts.field.empty()) {
        if (filename.find("nvdb_" + opts.field + "_") == std::string::npos) {
          continue;
        }
      }
      
      files.push_back(entry.path().string());
    }
  } catch (const std::exception& e) {
    std::cerr << "Error reading directory: " << e.what() << std::endl;
    exit(1);
  }
  
  std::sort(files.begin(), files.end());
  return files;
}

#ifdef HS_HAVE_NANOVDB

void mergeGrids(const Options& opts) {
  auto files = findNVDBFiles(opts);
  
  if (files.empty()) {
    std::cerr << "ERROR: No .nvdb files found in " << opts.inputDir << std::endl;
    exit(1);
  }
  
  std::cout << "Found " << files.size() << " NanoVDB files to merge" << std::endl;
  
  if (opts.verbose) {
    for (size_t i = 0; i < std::min(size_t(5), files.size()); i++) {
      std::cout << "  - " << fs::path(files[i]).filename().string() << std::endl;
    }
    if (files.size() > 5) {
      std::cout << "  ... and " << (files.size() - 5) << " more" << std::endl;
    }
  }
  
  std::cout << "\nMerging grids..." << std::endl;
  
  // Create a new grid builder for the merged result
  nanovdb::GridBuilder<float> mergedBuilder(0.0f);
  auto mergedAccessor = mergedBuilder.getAccessor();
  
  size_t totalActiveVoxels = 0;
  double voxelSize = 0.075;  // Will be read from first grid
  nanovdb::Vec3d worldOrigin(0, 0, 0);
  
  // Read and merge each grid
  for (size_t i = 0; i < files.size(); i++) {
    try {
      // Read grid from file
      auto handle = nanovdb::io::readGrid(files[i], "density");
      if (!handle) {
        std::cerr << "  WARNING: Failed to read " << fs::path(files[i]).filename().string() << std::endl;
        continue;
      }
      
      auto* grid = handle.grid<float>();
      if (!grid) {
        std::cerr << "  WARNING: Invalid grid in " << fs::path(files[i]).filename().string() << std::endl;
        continue;
      }
      
      // Get voxel size and origin from first grid
      if (i == 0) {
        voxelSize = grid->voxelSize()[0];
        worldOrigin = grid->worldBBox().min();
      }
      
      // Iterate over active voxels and copy to merged grid
      auto accessor = grid->getAccessor();
      size_t blockActiveCount = 0;
      
      // Get the bounding box and iterate over all coords
      auto bbox = grid->indexBBox();
      for (int z = bbox.min()[2]; z <= bbox.max()[2]; z++) {
        for (int y = bbox.min()[1]; y <= bbox.max()[1]; y++) {
          for (int x = bbox.min()[0]; x <= bbox.max()[0]; x++) {
            nanovdb::Coord ijk(x, y, z);
            if (grid->tree().isActive(ijk)) {
              float value = accessor.getValue(ijk);
              mergedAccessor.setValue(ijk, value);
              blockActiveCount++;
            }
          }
        }
      }
      
      totalActiveVoxels += blockActiveCount;
      
      if (opts.verbose && (i + 1) % 10 == 0) {
        std::cout << "  Merged " << (i + 1) << "/" << files.size() 
                  << " grids (" << totalActiveVoxels << " active voxels)" << std::endl;
      }
      
    } catch (const std::exception& e) {
      std::cerr << "  ERROR processing " << fs::path(files[i]).filename().string() 
                << ": " << e.what() << std::endl;
      continue;
    }
  }
  
  std::cout << "\nBuilding final merged grid..." << std::endl;
  auto finalHandle = mergedBuilder.getHandle<>(voxelSize, worldOrigin, "density");
  auto* finalGrid = finalHandle.grid<float>();
  
  std::cout << "Merged grid statistics:" << std::endl;
  std::cout << "  Active voxels: " << finalGrid->activeVoxelCount() << std::endl;
  
  auto idxBBox = finalGrid->indexBBox();
  std::cout << "  Index bbox: [" << idxBBox.min()[0] << "," << idxBBox.min()[1] << "," << idxBBox.min()[2] 
            << "] to [" << idxBBox.max()[0] << "," << idxBBox.max()[1] << "," << idxBBox.max()[2] << "]" << std::endl;
  
  auto wBBox = finalGrid->worldBBox();
  std::cout << "  World bbox: [" << wBBox.min()[0] << "," << wBBox.min()[1] << "," << wBBox.min()[2] 
            << "] to [" << wBBox.max()[0] << "," << wBBox.max()[1] << "," << wBBox.max()[2] << "]" << std::endl;
  
  // Write merged grid
  std::cout << "\nWriting merged grid to: " << opts.outputFile << std::endl;
  nanovdb::io::writeGrid(opts.outputFile, finalHandle);
  
  size_t fileSize = fs::file_size(opts.outputFile);
  std::cout << "  Output file size: " << (fileSize / (1024.0 * 1024.0)) << " MB" << std::endl;
  std::cout << "✅ Merge complete!" << std::endl;
}

#else

void mergeGrids(const Options& opts) {
  std::cerr << "ERROR: NanoVDB support not compiled in" << std::endl;
  std::cerr << "Install OpenVDB 8.0+ and reconfigure CMake with -DHS_HAVE_NANOVDB=ON" << std::endl;
  exit(1);
}

#endif

int main(int argc, char** argv) {
  Options opts = parseArgs(argc, argv);
  
  if (!fs::exists(opts.inputDir)) {
    std::cerr << "ERROR: Input directory not found: " << opts.inputDir << std::endl;
    return 1;
  }
  
  // Ensure output directory exists
  fs::path outPath(opts.outputFile);
  if (outPath.has_parent_path()) {
    fs::create_directories(outPath.parent_path());
  }
  
  mergeGrids(opts);
  
  return 0;
}

File: tools/merge_nanovdb.py
#!/usr/bin/env python3
"""
Merge multiple NanoVDB grid files from distributed simulation into a single grid.

Usage:
    python merge_nanovdb.py <input_dir> <output_file> [--timestep TIMESTEP]
    
Example:
    python merge_nanovdb.py /path/to/nvdb_files merged_lambda2_004320.nvdb --timestep 004320
"""

import sys
import os
import glob
import json
import argparse
from pathlib import Path

try:
    import pyopenvdb as vdb
    import numpy as np
except ImportError as e:
    print(f"ERROR: Required Python packages not found: {e}")
    print("Install with: pip install pyopenvdb numpy")
    sys.exit(1)


def parse_args():
    parser = argparse.ArgumentParser(
        description='Merge distributed NanoVDB files into a single grid',
        formatter_class=argparse.RawDescriptionHelpFormatter,
        epilog="""
Examples:
  # Merge all .nvdb files in a directory
  %(prog)s /data/nvdb_output merged.vdb
  
  # Merge only files from a specific timestep
  %(prog)s /data/nvdb_output merged_004320.vdb --timestep 004320
  
  # Merge with pattern matching
  %(prog)s /data/nvdb_output merged_lambda2.vdb --pattern "nvdb_lambda2_*.nvdb"
        """
    )
    parser.add_argument('input_dir', type=str,
                        help='Directory containing .nvdb files and .json metadata')
    parser.add_argument('output_file', type=str,
                        help='Output merged VDB file path')
    parser.add_argument('--timestep', type=str, default=None,
                        help='Only merge files from specific timestep (e.g., 004320)')
    parser.add_argument('--pattern', type=str, default="*.nvdb",
                        help='File pattern to match (default: *.nvdb)')
    parser.add_argument('--field', type=str, default=None,
                        help='Only merge files for specific field name')
    parser.add_argument('--verbose', '-v', action='store_true',
                        help='Print detailed progress information')
    return parser.parse_args()


def load_metadata(nvdb_file):
    """Load metadata JSON file associated with a .nvdb file"""
    json_file = nvdb_file + ".json"
    if not os.path.exists(json_file):
        return None
    
    with open(json_file, 'r') as f:
        return json.load(f)


def merge_grids(input_dir, output_file, timestep=None, pattern="*.nvdb", field=None, verbose=False):
    """Merge NanoVDB grids from multiple files"""
    
    # Find all matching .nvdb files
    search_pattern = os.path.join(input_dir, pattern)
    nvdb_files = glob.glob(search_pattern)
    
    if not nvdb_files:
        print(f"ERROR: No .nvdb files found matching: {search_pattern}")
        return False
    
    # Filter by timestep if specified
    if timestep:
        nvdb_files = [f for f in nvdb_files if f"_{timestep}." in f]
        if not nvdb_files:
            print(f"ERROR: No files found for timestep {timestep}")
            return False
    
    # Filter by field name if specified
    if field:
        nvdb_files = [f for f in nvdb_files if f"nvdb_{field}_" in os.path.basename(f)]
        if not nvdb_files:
            print(f"ERROR: No files found for field {field}")
            return False
    
    nvdb_files = sorted(nvdb_files)
    print(f"Found {len(nvdb_files)} NanoVDB files to merge")
    
    if verbose:
        for f in nvdb_files[:5]:
            print(f"  - {os.path.basename(f)}")
        if len(nvdb_files) > 5:
            print(f"  ... and {len(nvdb_files) - 5} more")
    
    # Load metadata from first file to get grid properties
    first_meta = load_metadata(nvdb_files[0])
    if first_meta:
        print(f"\nGrid properties:")
        print(f"  Field: {first_meta['fieldName']}")
        print(f"  Spacing: {first_meta['gridSpacing']}")
        print(f"  Threshold: {first_meta['threshold']}")
    
    # Create merged grid
    # Note: Since NanoVDB files already contain global coordinates,
    # we just need to read and combine them
    print(f"\nMerging grids...")
    
    try:
        # Read first grid to initialize
        merged_grid = vdb.read(nvdb_files[0], "density")
        total_active = merged_grid.activeVoxelCount()
        
        if verbose:
            print(f"  Loaded: {os.path.basename(nvdb_files[0])} ({merged_grid.activeVoxelCount()} active voxels)")
        
        # Merge remaining grids
        for i, nvdb_file in enumerate(nvdb_files[1:], 1):
            grid = vdb.read(nvdb_file, "density")
            
            # Combine grids (this should preserve the global coordinates)
            merged_grid.merge(grid)
            
            total_active += grid.activeVoxelCount()
            
            if verbose and i % 10 == 0:
                print(f"  Merged {i}/{len(nvdb_files)-1} grids... ({merged_grid.activeVoxelCount()} active)")
        
        print(f"\nMerged grid statistics:")
        print(f"  Total active voxels: {merged_grid.activeVoxelCount()}")
        print(f"  Input voxels (sum): {total_active}")
        print(f"  Bounding box: {merged_grid.evalActiveVoxelBoundingBox()}")
        
        # Write merged grid
        print(f"\nWriting merged grid to: {output_file}")
        vdb.write(output_file, grids=[merged_grid])
        
        file_size_mb = os.path.getsize(output_file) / (1024 * 1024)
        print(f"  Output file size: {file_size_mb:.2f} MB")
        print("✅ Merge complete!")
        
        return True
        
    except Exception as e:
        print(f"ERROR during merge: {e}")
        import traceback
        traceback.print_exc()
        return False


def main():
    args = parse_args()
    
    if not os.path.isdir(args.input_dir):
        print(f"ERROR: Input directory not found: {args.input_dir}")
        sys.exit(1)
    
    # Ensure output directory exists
    output_dir = os.path.dirname(args.output_file)
    if output_dir and not os.path.exists(output_dir):
        os.makedirs(output_dir)
    
    success = merge_grids(
        args.input_dir,
        args.output_file,
        timestep=args.timestep,
        pattern=args.pattern,
        field=args.field,
        verbose=args.verbose
    )
    
    sys.exit(0 if success else 1)


if __name__ == "__main__":
    main()

